File: JS/Tree/145. Binary Tree Postorder Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function(root) {  // Recursive
    const result = [];
    (function postorder(node) {
        if (node) {
            postorder(node.left);
            postorder(node.right);
            result.push(node.val);
        }
    })(root);
    return result;
};

var postorderTraversal = function(root) {  // Every time push the node twice so after we pop the node out, we can still have chance to visit it [Post Order]
    const result = [];
    const stack = [root, root];
    while(stack.length) {
        const node = stack.pop();
        if (node) {
            if (stack.length && node === stack[stack.length - 1]) { // If the current node equals to the one on top of the stack, it means we are traversing to left node
                stack.push(node.right);
                stack.push(node.right);
                stack.push(node.left);
                stack.push(node.left);
            } else {
                result.push(node.val);  // if the current node doesn't equal to the one on top of the stack, it means we are backing from the right subtree traversing
            }
        }
    }
    return result;
};

File: Golang/Tree/145.binary-tree-postorder-traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func postorderTraversalRecursive(root *TreeNode) []int {
	ret := []int{}
	var recursive func(node *TreeNode)
	recursive = func(node *TreeNode) {
		if node != nil {
			recursive(node.Left)
			recursive(node.Right)
			ret = append(ret, node.Val)
		}
	}
	recursive(root)
	return ret
}

func postorderTraversalPointer(root *TreeNode) []int {
	ret := []int{}
	if root != nil {

		stack := []*TreeNode{root}
		for len(stack) > 0 {
			node := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if node.Left != nil {
				stack = append(stack, node.Left)
			}
			if node.Right != nil {
				stack = append(stack, node.Right)
			}
			ret = append(ret, node.Val)
		}
	}
	return ret
}
